src/server/snap.rs

```diff
@@ -238,41 +238,21 @@ impl<R: RaftStoreRouter + 'static> Runnable<Task> for Runner<R> {
             Task::Write(token, mut data) => {
                 SNAP_TASK_COUNTER.with_label_values(&["write"]).inc();
                 match self.files.entry(token) {
-                    Entry::Occupied(mut e) => {
-                        let mut finish = true;
-                        {
-                            let (left, right) = data.slice();
-                            let limiter = self.snap_mgr.get_limiter();
-                            if let Err(err) = limiter.write(&mut e.get_mut().0, left) {
-                                error!(
-                                    "failed to write data to snapshot file {} for token {:?}: {:?}",
-                                    e.get_mut().0.path(),
-                                    token,
-                                    err
-                                );
-                                let (_, msg) = e.remove();
-                                let key =
-                                    SnapKey::from_snap(msg.get_message().get_snapshot()).unwrap();
-                                self.snap_mgr.deregister(&key, &SnapEntry::Receiving);
-                                finish = false;
-                            } else if let Err(err) = limiter.write(&mut e.get_mut().0, right) {
-                                error!(
-                                    "failed to write data to snapshot file {} for token {:?}: {:?}",
-                                    e.get_mut().0.path(),
-                                    token,
-                                    err
-                                );
-                                let (_, msg) = e.remove();
-                                let key =
-                                    SnapKey::from_snap(msg.get_message().get_snapshot()).unwrap();
-                                self.snap_mgr.deregister(&key, &SnapEntry::Receiving);
-                                finish = false;
-                            }
-                        }
-                        if finish {
-                            data.set_empty();
-                        }
-                    }
+                    Entry::Occupied(mut e) => if let Err(err) =
+                        data.write_all_to(&mut LimiterWriter {
```

Comment:
create the writer first, then pass to the function here.

Reply:
I tried that, but in that case we will have two mutable reference `e.get_mut().0`, it doesn't work.